balloon/src/lib.rs

```diff
@@ -0,0 +1,347 @@
+                    digest.update(&cnt.to_le_bytes());
+                    cnt = cnt.wrapping_add(1);
+                    digest.update(salt);
+                    digest.update(&u32::try_from(t).unwrap().to_le_bytes());
+                    digest.update(&u32::try_from(m).unwrap().to_le_bytes());
+                    digest.update(&i.to_le_bytes());
+                    let other =
+                        u32::from_le_bytes(digest.finalize_reset()[..4].try_into().unwrap())
+                            % self.params.s_cost.get();
+                    let other = usize::try_from(other).unwrap();
+
+                    // buf[m] = hash(cnt++, buf[m], buf[other])
+                    digest.update(&cnt.to_le_bytes());
+                    cnt = cnt.wrapping_add(1);
+                    digest.update(&memory_blocks[m]);
+                    digest.update(&memory_blocks[other]);
+                    memory_blocks[m] = digest.finalize_reset();
+                }
+            }
+        }
+
+        // Step 3. Extract output from buffer.
+        // return buf[s_cost-1]
+        Ok(memory_blocks[s_cost - 1].clone())
+    }
+}
+
+#[cfg(all(feature = "alloc", feature = "password-hash"))]
+#[cfg_attr(docsrs, doc(cfg(feature = "alloc")))]
+#[cfg_attr(docsrs, doc(cfg(feature = "password-hash")))]
+impl<D: Digest> PasswordHasher for Balloon<'_, D> {
+    type Params = Params;
+
+    fn hash_password<'a, S>(
+        &self,
+        password: &[u8],
+        salt: &'a S,
+    ) -> password_hash::Result<PasswordHash<'a>>
+    where
+        S: AsRef<str> + ?Sized,
+    {
+        let salt = Salt::try_from(salt.as_ref())?;
+        let mut salt_arr = [0u8; 64];
+        let salt_bytes = salt.b64_decode(&mut salt_arr)?;
+
+        let output = password_hash::Output::new(&self.hash(password, salt_bytes)?)?;
+
+        Ok(PasswordHash {
+            algorithm: Ident::new("balloon"),
+            version: Some(1),
+            params: ParamsString::try_from(&self.params)?,
+            salt: Some(salt),
+            hash: Some(output),
+        })
+    }
+
+    fn hash_password_customized<'a>(
+        &self,
+        password: &[u8],
+        alg_id: Option<Ident<'a>>,
+        version: Option<Decimal>,
+        params: Params,
+        salt: impl Into<Salt<'a>>,
+    ) -> password_hash::Result<PasswordHash<'a>> {
+        if let Some(alg_id) = alg_id {
+            if alg_id.as_str() != "balloon" {
+                return Err(password_hash::Error::Algorithm);
+            }
+        }
+
+        if let Some(version) = version {
+            if version != 1 {
+                return Err(password_hash::Error::Version);
+            }
+        }
+
+        let salt = salt.into();
+
+        Self::new(params, self.secret).hash_password(password, salt.as_str())
+    }
+}
+
+#[cfg(all(test, feature = "alloc", feature = "password-hash"))]
+#[test]
+fn hash_simple_retains_configured_params() {
```

Comment:
I'm happy to add any other tests, but there are no test vectors out there as far as I could find.

Potentially I could compile the reference implementation as a dev-dependency (I don't know how much work that would be), but for that to be useful, it would require the implementation details to be exactly the same, which it already isn't, see all the notes above.

Reply:
Agreed, this is a tough call. It's definitely risky to publish something that has no test vectors to speak of. @tarcieri what do you think?

Reply:
If there aren't test vectors to speak of, then at the very least I would suggest creating some with the reference implementation and leave the tool used to do so online somewhere.

If you could make a Rust FFI crate to so, and check it in as part of the initial implementation, that'd be great. We can potentially remove it from the source tree later, but it'd be nice to capture the provenance of the test vectors in a way that's reproducible and exists in the source code history.

Reply:
I don't mind, but that would require us to properly mirror the prototype implementation, which would lead us back to https://github.com/RustCrypto/password-hashes/pull/232#discussion_r700919220 and `EVP_EncryptUpdate` which is still have little clue what it does, but if it involves some sort of encryption, I think this isn't what we want, right?

Reply:
`EVP_EncryptUpdate` is a (now deprecated) OpenSSL API:

https://www.openssl.org/docs/manmaster/man3/EVP_EncryptUpdate.html

You said it was doing something with AES... is it possible it was deriving an AES key?

The test vectors should probably cover the raw output. Worst case you could add a public API for that so you don't have to go through an OpenSSL compatibility layer.

Mirroring the sources is fine if they aren't too large. It'd just be nice to have a safe place where the test vectors can be generated/regenerated.

Reply:
Sure, there are still other things I will have to make sure are exactly the same, I will make an attempt.

Reply:
I had a branch working on it, but I guess our newest decision is to not try and mirror the prototype implementation, so I will close this. If we change our mind we can still revive it. The FFI crate is here: https://github.com/daxpedda/balloon-proto.